src/compute/src/communication.rs

```diff
@@ -68,46 +69,128 @@ fn create_sockets(
+///
+/// If an EOF or an error other than those two is signaled, the connection is broken
+/// and should be set to `None`.
+fn gc_broken_connections<'a, I>(conns: I, first_idx: usize)
+where
+    I: IntoIterator<Item = &'a mut Option<TcpStream>>,
+{
+    let mut buf = [0];
+    for (i, maybe_conn) in conns.into_iter().enumerate() {
+        trace!("peeking {} to detect whether it's broken", first_idx + i);
+        if let Some(conn) = maybe_conn {
+            let closed = match conn.peek(&mut buf) {
+                Ok(0) => true, // EOF
+                Ok(_) => false,
+                Err(err) => err.kind() != std::io::ErrorKind::WouldBlock,
+            };
+            if closed {
+                warn!(
+                    "Connection {} has gone down; we will try to establish it again.",
+                    i + first_idx
+                );
+                *maybe_conn = None;
+            } else {
+                trace!("Peek OK")
+            }
+        }
+    }
+}
+
 /// Result contains connections [0, my_index - 1].
 fn start_connections(
     addresses: Arc<Vec<String>>,
     my_index: usize,
 ) -> Result<Vec<Option<TcpStream>>, io::Error> {
-    let results = addresses
-        .iter()
-        .take(my_index)
-        .enumerate()
-        .map(|(index, address)| loop {
-            match TcpStream::connect(address) {
-                Ok(mut stream) => {
-                    stream.set_nodelay(true).expect("set_nodelay call failed");
-                    stream
-                        .write_all(&my_index.to_ne_bytes())
+    let addresses = loop {
+        match addresses
+            .iter()
+            .take(my_index)
+            .map(|address| {
+                address
+                    .as_str()
+                    .to_socket_addrs()
+                    .map(|addrs| addrs.into_first()) // `man getaddrinfo` claims it returns at least one element
+                    .map_err(|e| (e, address.clone()))
+            })
+            .collect::<Result<Vec<_>, _>>()
+        {
+            Ok(addresses) => break addresses,
+            Err((e, addr)) => {
+                warn!("Failed to resolve {addr}: {e}, will retry");
+                sleep(Duration::from_secs(1));
+            }
+        }
+    };
+    let mut results: Vec<_> = (0..my_index).map(|_| None).collect();
+
+    // We do not want to provide opportunities for the startup
+    // sequence to hang waiting for one of its peers, not noticing that another has gone down;
+    // empirically, we have found it difficult to reason
+    // about whether this results in crash loops and global deadlocks.
+    // Thus, every second, check whether a previously-established connection has
+    // gone away; if so, we clear it and retry it again later.
+    let mut i = 0;
+    while results.iter().any(|r| r.is_none()) {
+        if results[i].is_none() {
+            match TcpStream::connect_timeout(&addresses[i], Duration::from_secs(1)) {
```

Comment:
This seems a little dangerous. It means we'll never be able to connect if connections take more than 1s, because we restart the connection from scratch on each turn of the loop. Maybe that's okay, but I wouldn't want to bet on it. ISTM you want a separate thread that's doing the GC. Alternatively it seems it could be easier to express this logic with futures? You could `select!` on the connection future and a `time::sleep(Duration::from_secs(1))`, do the GC check, and then come back and resume `await`ing on the _same_ connection attempt.

Reply:
If TCP connections are routinely taking longer than 1s to spin up then something is badly messed up anyway… but fair point, we might as well deal with this case since it’s not too hard to do so.

Reply:
I would rather use a separate thread than introduce async/futures stuff where it’s not already being used, so I’ll go with that option.

Reply:
Oh, actually I don't think we need to do either. It should be fine to just use `connect` without a timeout, here. The important thing isn't the 1s timeout, it's clearing old broken connections and retrying them.

So I think the code can be left 99% as-is, and I can just change `connect_timeout` back to `connect`.